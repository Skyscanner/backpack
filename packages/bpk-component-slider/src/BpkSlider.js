/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import Slider from 'react-slider';
import PropTypes from 'prop-types';

import { cssModules, isRTL } from '../../bpk-react-utils';

import STYLES from './BpkSlider.module.scss';

const getClassName = cssModules(STYLES);

const BpkSlider = (props) => {
  const { className, large, ...rest } = props;
  const invert = isRTL();
  const classNames = [getClassName('bpk-slider')];
  const thumbClassNames = [getClassName('bpk-slider__handle')];
  const trackClassNames = [getClassName('bpk-slider__bar')];

  const isRange = (rest.value || rest.defaultValue || []).length > 1;

  if (isRange) {
    classNames.push(getClassName('bpk-slider--range'));
  }
  if (className) {
    classNames.push(getClassName(className));
  }
  if (large) {
    classNames.push(getClassName('bpk-slider--large'));
    thumbClassNames.push(getClassName('bpk-slider__handle--large'));
  }

  return (
    <Slider
      {...rest}
      withTracks
      invert={invert}
      className={classNames.join(' ')}
      thumbClassName={thumbClassNames.join(' ')}
      thumbActiveClassName={getClassName('bpk-slider__handle--active')}
      trackClassName={trackClassNames.join(' ')}
    />
  );
};

BpkSlider.propTypes = {
  className: PropTypes.string,
  large: PropTypes.bool,
};

BpkSlider.defaultProps = {
  className: null,
  large: false,
};

export default BpkSlider;
