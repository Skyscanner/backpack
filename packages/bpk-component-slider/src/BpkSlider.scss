/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2019 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@import '~bpk-mixins';

$slider-control-inset-size: 20 * $bpk-one-pixel-rem;

.bpk-slider {
  height: $bpk-spacing-lg;
  -webkit-tap-highlight-color: transparent;

  &--large {
    height: $bpk-spacing-xxl;
  }

  &__handle {
    display: flex;
    width: $bpk-spacing-lg;
    height: $bpk-spacing-lg;
    justify-content: center;
    align-items: center;
    border: $bpk-one-pixel-rem solid;
    border-radius: $bpk-border-radius-pill;
    background-color: $bpk-color-white;
    cursor: pointer;

    @include bpk-apply-gray-100-to(border-color);
    @include bpk-box-shadow-sm;

    &::after {
      content: '';
      display: block;
      width: $slider-control-inset-size;
      height: $slider-control-inset-size;
      border-radius: $bpk-border-radius-pill;
      box-shadow: inset 0 0 0 $bpk-one-pixel-rem $bpk-color-gray-200;
    }

    &--active {
      transform: scale(1.2) translate3d(0, 0, 0);
    }

    &--large {
      width: $bpk-spacing-xxl;
      height: $bpk-spacing-xxl;
      border-radius: $bpk-border-radius-pill-lg;

      &::after {
        width: $bpk-spacing-lg;
        height: $bpk-spacing-lg;
        border-radius: $bpk-border-radius-pill-lg;
      }
    }
  }

  &__bar {
    top: calc(50% - (#{$bpk-spacing-xs} / 2));
    height: $bpk-spacing-xs;
    border-radius: $bpk-border-radius-sm;

    @include bpk-apply-gray-100-to(background-color);
  }

  // stylelint-disable-next-line selector-attribute-quotes
  > div[class$='0'] {
    @include bpk-themeable-property(background-color, --bpk-slider-bar-color, $bpk-color-blue-500);
  }

  &--range {
    // stylelint-disable-next-line selector-attribute-quotes
    > div[class$='0'] {
      @include bpk-apply-gray-100-to(background-color);
    }
    // stylelint-disable-next-line selector-attribute-quotes
    > div[class$='1'] {
      @include bpk-themeable-property(background-color, --bpk-slider-bar-color, $bpk-color-blue-500);
    }
  }
}
