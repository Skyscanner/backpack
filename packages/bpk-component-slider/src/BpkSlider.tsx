/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as Slider from '@radix-ui/react-slider';

import { cssModules, isRTL } from '../../bpk-react-utils';

import STYLES from './BpkSlider.module.scss';

const getClassName = cssModules(STYLES);

export type Props = {
  max: number;
  min: number;
  minDistance?: number;
  step?: number;
  onChange: (value: number[] | number) => void;
  onAfterChange?: (value: number[] | number) => void;
  value: number[] | number;
  ariaLabel: string[];
  ariaValuetext?: string[];
  [rest: string]: any;
};

const BpkSlider = ({
  ariaLabel,
  ariaValuetext,
  max,
  min,
  minDistance,
  onAfterChange,
  onChange,
  step,
  value,
  ...rest
}: Props) => {
  const invert = isRTL();
  const currentValue = Array.isArray(value) ? value : [value];

  const processSliderValues = (
    sliderValues: number[],
    callback?: (val: number | number[]) => void,
  ) => {
    const val = sliderValues.length === 1 ? sliderValues[0] : sliderValues;
    if (callback) {
      callback(val);
    }
  };

  const handleOnChange = (sliderValues: number[]) => {
    processSliderValues(sliderValues, onChange);
  };

  const handleOnAfterChange = (sliderValues: number[]) => {
    processSliderValues(sliderValues, onAfterChange);
  };

  return (
    <Slider.Root
      className={getClassName('bpk-slider')}
      defaultValue={currentValue}
      min={min}
      max={max}
      step={step || 1}
      onValueChange={handleOnChange}
      onValueCommit={handleOnAfterChange}
      inverted={invert}
      minStepsBetweenThumbs={minDistance}
      {...rest}
    >
      <Slider.Track className={getClassName('bpk-slider__track')}>
        <Slider.Range className={getClassName('bpk-slider__range')} />
      </Slider.Track>
      {currentValue.map((val, index) => (
        <Slider.Thumb
          key={ariaLabel[index]}
          aria-label={ariaLabel[index]}
          aria-valuetext={ariaValuetext ? ariaValuetext[index] : val.toString()}
          className={getClassName('bpk-slider__thumb')}
          aria-valuenow={currentValue[index]}
        />
      ))}
    </Slider.Root>
  );
};

export default BpkSlider;
