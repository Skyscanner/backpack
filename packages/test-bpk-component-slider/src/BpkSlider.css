@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-slider{height:1.875rem;-webkit-tap-highlight-color:transparent}.bpk-slider--large{height:2.625rem}.bpk-slider__handle{display:flex;width:1.875rem;height:1.875rem;justify-content:center;align-items:center;border:.0625rem solid #dddde5;border-radius:1.125rem;background-color:#fff;cursor:pointer;box-shadow:0px 1px 3px 0px rgba(37,32,31,0.3)}.bpk-slider__handle::after{content:'';display:block;width:1.25rem;height:1.25rem;border-radius:1.125rem;box-shadow:inset 0 0 0 .0625rem #cdcdd7}.bpk-slider__handle--active{transform:scale(1.2) translate3d(0, 0, 0)}.bpk-slider__handle--large{width:2.625rem;height:2.625rem;border-radius:1.3125rem}.bpk-slider__handle--large::after{width:1.875rem;height:1.875rem;border-radius:1.3125rem}.bpk-slider__bar{top:calc(50% - (.375rem / 2));height:.375rem;border-radius:.375rem;background-color:#dddde5}.bpk-slider>div[class$='0']{background-color:#02122c;background-color:var(--bpk-slider-bar-color, #02122c)}.bpk-slider--range>div[class$='0']{background-color:#dddde5}.bpk-slider--range>div[class$='1']{background-color:#02122c;background-color:var(--bpk-slider-bar-color, #02122c)}
