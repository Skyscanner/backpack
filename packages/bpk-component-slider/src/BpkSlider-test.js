/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import BpkSlider from './BpkSlider';

jest.mock('react-slider', () => 'react-slider');

describe('BpkSlider', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<BpkSlider min={0} max={100} value={25} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a "className" attribute', () => {
    const tree = renderer
      .create(<BpkSlider min={0} max={100} value={25} className="my-slider" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a "large" attribute', () => {
    const tree = renderer
      .create(<BpkSlider min={0} max={9} value={2} large />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a "step" attribute', () => {
    const tree = renderer
      .create(<BpkSlider min={0} max={100} value={2} step={10} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a range of values', () => {
    const tree = renderer
      .create(<BpkSlider min={0} max={100} value={[10, 90]} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a minimum distance between controls', () => {
    const tree = renderer
      .create(<BpkSlider min={0} max={100} value={[10, 90]} minDistance={20} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
