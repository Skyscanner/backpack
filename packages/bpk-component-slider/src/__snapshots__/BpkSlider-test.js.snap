// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkSlider should render correctly 1`] = `
<div
  className="bpk-slider"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="bpk-slider__bar bpk-slider__bar-0"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-1"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={25}
    className="bpk-slider__handle bpk-slider__handle-0 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  />
</div>
`;

exports[`BpkSlider should render correctly with a "className" attribute 1`] = `
<div
  className="bpk-slider my-slider"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="bpk-slider__bar bpk-slider__bar-0"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-1"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={25}
    className="bpk-slider__handle bpk-slider__handle-0 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  />
</div>
`;

exports[`BpkSlider should render correctly with a "large" attribute 1`] = `
<div
  className="bpk-slider bpk-slider--large"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="bpk-slider__bar bpk-slider__bar-0"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-1"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    aria-valuemax={9}
    aria-valuemin={0}
    aria-valuenow={2}
    className="bpk-slider__handle bpk-slider__handle--large bpk-slider__handle bpk-slider__handle--large-0 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  />
</div>
`;

exports[`BpkSlider should render correctly with a "step" attribute 1`] = `
<div
  className="bpk-slider"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="bpk-slider__bar bpk-slider__bar-0"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-1"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={0}
    className="bpk-slider__handle bpk-slider__handle-0 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  />
</div>
`;

exports[`BpkSlider should render correctly with a minimum distance between controls 1`] = `
<div
  className="bpk-slider bpk-slider--range"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="bpk-slider__bar bpk-slider__bar-0"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-1"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-2"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={10}
    className="bpk-slider__handle bpk-slider__handle-0 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  />
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={90}
    className="bpk-slider__handle bpk-slider__handle-1 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 2,
      }
    }
    tabIndex={0}
  />
</div>
`;

exports[`BpkSlider should render correctly with a range of values 1`] = `
<div
  className="bpk-slider bpk-slider--range"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="bpk-slider__bar bpk-slider__bar-0"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-1"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    className="bpk-slider__bar bpk-slider__bar-2"
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "willChange": "",
      }
    }
  />
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={10}
    className="bpk-slider__handle bpk-slider__handle-0 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 1,
      }
    }
    tabIndex={0}
  />
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={90}
    className="bpk-slider__handle bpk-slider__handle-1 "
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "0px",
        "position": "absolute",
        "willChange": "",
        "zIndex": 2,
      }
    }
    tabIndex={0}
  />
</div>
`;
